confmap/converter/templateconverter/converter.go

```diff
@@ -0,0 +1,262 @@
+		if !ok {
+			return fmt.Errorf("'templates::receivers::%s' must be a string", templateType)
+		}
+		parsedTemplate, err := template.New(templateType).Parse(templateStr)
+		if err != nil {
+			return err
+		}
+		c.receiverTemplates[templateType] = parsedTemplate
+	}
+	return nil
+}
+
+func (c *converter) expandTemplate(cfg *templateConfig, id *instanceID) {
+	// Delete the reference to this template instance and
+	// replace it with the rendered receiver(s).
+	delete(c.cfg["receivers"].(map[string]any), id.withPrefix("template"))
+	for receiverID, receiverCfg := range cfg.Receivers {
+		c.cfg["receivers"].(map[string]any)[receiverID] = receiverCfg
+	}
+
+	// Special case where the template only contains receivers. In this case,
+	// we can just substitute the rendered receivers in place of the template ID.
+	if len(cfg.Processors) == 0 && len(cfg.Pipelines) == 0 {
+		c.expandSimple(cfg, id)
+	} else {
+		c.expandComplex(cfg, id)
+	}
+}
+
+// Special case where the template only contains receivers. In this case,
+// we can just substitute the rendered receivers in place of the template ID.
+func (c *converter) expandSimple(cfg *templateConfig, id *instanceID) {
+	pipelinesMap := c.cfg["service"].(map[string]any)["pipelines"].(map[string]any)
+	for _, pipelineMap := range pipelinesMap {
+		receiverIDs := pipelineMap.(map[string]any)["receivers"].([]any)
+		newReceiverIDs := make([]any, 0, len(receiverIDs))
+		for _, receiverID := range receiverIDs {
+			if receiverID != id.withPrefix("template") {
+				newReceiverIDs = append(newReceiverIDs, receiverID)
+			}
+		}
+
+		if len(newReceiverIDs) == len(receiverIDs) {
+			// This pipeline did not use the template
+			continue
+		}
+
+		for receiverID := range cfg.Receivers {
+			newReceiverIDs = append(newReceiverIDs, receiverID)
+		}
+		// This makes tests deterministic
+		sort.Slice(newReceiverIDs, func(i, j int) bool {
+			return newReceiverIDs[i].(string) < newReceiverIDs[j].(string)
+		})
+		pipelineMap.(map[string]any)["receivers"] = newReceiverIDs
+	}
+}
+
+// Any partial pipelines defined in a template will be expanded into full pipelines.
+// These will all use a forward connector to emit data to the pipelines in which
+// the template was used as a receiver.
+func (c *converter) expandComplex(cfg *templateConfig, id *instanceID) {
+	pipelinesMap := c.cfg["service"].(map[string]any)["pipelines"].(map[string]any)
+	// Update the processors section by adding any rendered processors.
+	if len(cfg.Processors) > 0 && c.cfg["processors"] == nil {
+		c.cfg["processors"] = make(map[string]any, len(cfg.Processors))
+	}
+	for processorID, processorCfg := range cfg.Processors {
+		c.cfg["processors"].(map[string]any)[processorID] = processorCfg
+	}
+
+	// Add a dedicated forward connector for this template instance.
+	// This will consume data from the partial pipelines and emit to
+	// any top-level pipelines that used the template as a receiver.
+	connectorID := id.withPrefix("forward")
```

Comment:
Do we have the ability to validate that the forward connector is available here? I realize it's outside the responsibility of a converter to validate the config this way, but I worry about the case where a user of a custom distribution attempts to use a template and receives an error saying that the forward connector isn't a known component when they never specified it in any of their configs.

Alternatively, could we check if it's available somewhere in `otelcol` if a template is provided?

Reply:
This is a good question. I haven't found a good way to do this. Any more specific ideas?

Generally I agree it would be nice to have a better error message but maybe documentation would be sufficient.

Reply:
I think it will have to just be documentation for now. Any code aware of the available components only deals with a resolved config that has already had the `templates` section removed, so there's no reliable way to determine if templates were used.